src/geodesy/ntv2_grid.rs

```diff
@@ -0,0 +1,144 @@
+) -> Result<(Vec<f64>, Vec<Vec<[f64; 2]>>)> {
+    let lat_0 = view.get_float64_endian(offset + NTV2_SUBGRID_NLAT, is_le); // Latitude of the first (typically northernmost) row of the grid
+    let lat_1 = view.get_float64_endian(offset + NTV2_SUBGRID_SLAT, is_le); // Latitude of the last (typically southernmost) row of the grid
+
+    let lon_0 = view.get_float64_endian(offset + NTV2_SUBGRID_WLON, is_le); // Longitude of the first (typically westernmost) column of each row
+    let lon_1 = view.get_float64_endian(offset + NTV2_SUBGRID_ELON, is_le); // Longitude of the last (typically easternmost) column of each row
+
+    let dlat = view.get_float64_endian(offset + NTV2_SUBGRID_DLAT, is_le); // Signed distance between two consecutive rows
+    let dlon = view.get_float64_endian(offset + NTV2_SUBGRID_DLON, is_le); // Signed distance between two consecutive columns
+
+    let num_nodes = view.get_int32_endian(offset + NTV2_SUBGRID_GSCOUNT, is_le) as usize;
+
+    let grid_start_offset = offset + NTV2_HEADER_SIZE;
+    let grid_end_offset = grid_start_offset + num_nodes * NTV2_NODE_SIZE;
+
+    if grid_end_offset > view.byte_length() {
+        return Err(Error::InvalidNtv2GridFormat("Invalid Grid: Too Short"));
+    }
+
+    // As defined by https://web.archive.org/web/20140127204822if_/http://www.mgs.gov.on.ca:80/stdprodconsume/groups/content/@mgs/@iandit/documents/resourcelist/stel02_047447.pdf (pg 30)
+    let rows = (((lat_1 - lat_0) / dlat).abs() + 1.0).floor() as usize;
+    let cols = (((lon_0 - lon_1) / dlon).abs() + 1.0).floor() as usize;
+
+    if num_nodes != (rows * cols) {
+        return Err(Error::InvalidNtv2GridFormat(
+            "Invalid Grid: Number of nodes does not match the grid size",
+        ));
+    }
+
+    // NTv2 nodes are order from SE to NW! Here we break them up into rows and columns
+    // reversing the order so we end up with a grid ordered from NW to SE.
+    //         NW=(r₀:c₀)      NE=(r₀:cₙ)           SW=(rₙ:c₀)      SE=(rₙ:cₙ)
+    //      [[[lat₀, lon₀]...[latₙ, lonₙ]]ᵣ₀,..., [[lat₀, lon₀]...[latₙ, lonₙ]]ᵣₙ]
+    //
+    // lat/lon values are in minutes-of-arc as is expected by the Gravsoft reader in Rust Geodesy
+    // [RG Says it's minutes of arc but it's dividing by 3600 so it's actually expecting seconds of arc?!](https://github.com/busstoptaktik/geodesy/blob/3bb3beec2593ad9c35ef3c7e6d92383e6c9828ed/src/grid/mod.rs#L182-L192)
+    let grid = (0..num_nodes)
+        .map(|i| {
+            let offset = grid_start_offset + i * NTV2_NODE_SIZE;
+            let lat_corr = view.get_float32_endian(offset + NTV2_NODE_LAT_CORRN, is_le);
+            let lon_corr = view.get_float32_endian(offset + NTV2_NODE_LON_CORRN, is_le);
+            [lat_corr.into(), lon_corr.into()]
+        })
+        .rev() // Because only the insane work SE to NW!
+        .collect::<Vec<[f64; 2]>>()
+        .chunks(cols) // Chunk into rows
+        .map(|chunk| chunk.to_vec())
+        .collect::<Vec<Vec<[f64; 2]>>>();
```

Comment:
From reading the[gravsoft reader in rust geodesy](https://github.com/busstoptaktik/geodesy/blob/3bb3beec2593ad9c35ef3c7e6d92383e6c9828ed/src/grid/mod.rs#L182-L192) it looks like it's expecting grid values in `seconds-of-arc` not `minutes-of-arc` as the comment suggests. 

@busstoptaktik are you able to confirm this?

After changing the values to seconds my control is still incorrect, but somewhat closer - at least in the northing?
```bash
Diff Expected Coords
364.1372725752972 0.024172578938305378 0
364.1377419260425 0.024571018293499947 0
```
Still not correct, but closer?!

Reply:
You are right that my comment in [the Gravsoft reader](https://github.com/busstoptaktik/geodesy/blob/3bb3beec2593ad9c35ef3c7e6d92383e6c9828ed/src/grid/mod.rs#L182-L192) is wrong.

Over at [Esri's GSB](https://github.com/Esri/ntv2-file-routines)-converter, which comes with a `README.md` which is quite a bit easier to read than the original NTv2 specification, you'll find these two nuggets:

1. The header includes this: `GS_TYPE   string  Grid-shift units: "SECONDS", "MINUTES", or "DEGREES"`
2. (item 3 of Esri's "Notes" section):
   - All published NTv2 files that we have seen have a GS_TYPE of "SECONDS". This has the advantage that the latitude, longitude, and increment values in the sub-file records are then usually whole numbers.

So it appears safe to assume that NTv2 files are in seconds-of-arc. Perhaps the difference is in
- the sign (should you add or subtract the correction),
- the transformation-direction of the grid (no iteration when it should, if using a grid converting in the "wrong" direction, "repaired" by switching signs), or
- the order of the corrections (dlat, dlon) vs. (dlon, dlat)?

Reply:
Agreed on the Esri README.md, much more helpful. 

Ah right, switching signs on the longitutde correction gets us most of the way there. I've now got a much closer diff between RG and PROJ of 

```bash
Diff Expected Coords
0.01048276152505423 0.024172578938305378 0
0.01068291185038106 0.024571018293499947 0
```

This make sense, as far as I can tell, because we've switched the direction of the rows. Confusingly we don't need to switch the sign on the latitude. But after stairing crossed at the picture on page 30 of the [official spec](https://web.archive.org/web/20140127204822if_/http://www.mgs.gov.on.ca:80/stdprodconsume/groups/content/@mgs/@iandit/documents/resourcelist/stel02_047447.pdf) I think by luck it's direction ends up being the same as the Gravsoft one. 

Would you expect the difference to be smaller than the above?

Reply:
@busstoptaktik I tried a few more things this morning to see if I could get the difference between PROJ and Geodesy down. 

- Re point one: Making the `lon` correction negative resulted in the above diff. 
- Re point two: To test directionality I added an `inv` to the gridshift step. If I also made the `lat` correction negative I get this diff:
```bash
Diff Expected Coords
0.030175553816661704 0.0053491536527872086 0
0.03037858445532038 0.005748012103140354 0
```
Which is better in the northing than the above but worse in the easting. Could you say more about how `"repaired" by switching signs` could fix it i.e without hacking the `inv` operator. It feels like a combination of switching signs and directionality could improve things?

- Re point three: No manner of switching `dlat` and `dlon` came anywhere as close as the first two attempts.

Reply:
> Could you say more about how "repaired" by switching signs could fix it i.e without hacking the inv operator

Spelled out in more detail: Could the situation be that

- Actually the grid represents a transformation from A to B, hence needing iteration when going from B to A
- You expect it to represent a transformation from B to A, and by juggling the signs you get "almost there", because the iterative correction for the inverse case is almost always quite small

Considering the odd sign conventions of NTv2 (and I'm not really sure whether there's even a difference between the sign of the longitude, and the sign of the longitude correction), it's really tricky to make it fit into a Gravsoft world, where:

- corrections and longitudes share the same sign, and
- if the grid is supposed to take you fra A to B, you get there by adding the grid values to your A coordinates, and
- if you need to go in the opposite direction, the code will flip the signs and take a few iterative steps